src/crystal/system/unix/event_io_uring.cr

```diff
@@ -1,43 +1,48 @@
-{% skip_file unless flag?(:linux) %}
-
 require "./io_uring"
 
 # :nodoc:
-struct Crystal::IoUringEvent < Crystal::Event
+struct Crystal::IoUring::Event < Crystal::EventLoop::Event
   enum Type
     Resume
     Timeout
     ReadableFd
     WritableFd
   end
 
-  def initialize(@io_uring : Crystal::System::IoUring, @type : Type, @fd : Int32, &callback : Int32 ->)
-    @callback = Box.box(callback)
+  def initialize(@io_uring : Crystal::System::IoUring, @type : Type, @fd : Int32 = -1, &@callback : Int32 ->)
+    @action_id = 0u64
   end
 
   def free : Nil
+    delete
   end
 
   def delete : Nil
+    return if @action_id == 0u64
+    @io_uring.delete_completion_action(@action_id)
     if @type.timeout?
-      @io_uring.timeout_remove(@callback)
+      @io_uring.submit_timeout_remove(@action_id)
     end
+    @action_id = 0u64
   end
 
   def add(timeout : Time::Span?) : Nil
+    delete
+    @action_id = @io_uring.register_completion_action(@callback)
+
     timeout = nil if timeout == Time::Span::ZERO
 
     case @type
     in .resume?, .timeout?
       if timeout
-        @io_uring.timeout(timeout, @callback)
+        @io_uring.submit_timeout(timeout, action_id: @action_id)
       else
-        @io_uring.nop(@callback)
+        @io_uring.submit_nop(action_id: @action_id)
       end
     in .readable_fd?
-      @io_uring.wait_readable(@fd, @callback, timeout: timeout)
+      @io_uring.submit_poll_add(@fd, Crystal::System::Syscall::POLLIN, action_id: @action_id, timeout: timeout)
     in .writable_fd?
-      @io_uring.wait_writable(@fd, @callback, timeout: timeout)
+      @io_uring.submit_poll_add(@fd, Crystal::System::Syscall::POLLOUT, action_id: @action_id, timeout: timeout)
```

Comment:
So hmm, if I understand this correctly, this make it so that instead of going through events and wait_read/writeable it just submits a poll instead? I see one issue with that and that is the case when several fibers wait on the same file descriptor - it will wake every fiber that has submitted a poll instead of just waking up the first fiber. This is a difference to how the non-uring fiber loop works and I'm not certain that behavior is something that is good enough as it may mean that when a fiber actually gets woken up the fd is no longer actionable as some other fiber may have acted on it.

Also known as "thundering herd problem"

Reply:
Right now the goal is to mimic the exact same behavior as libevent. This `action_id` there points to a callback (see `create_fd_write_event`). It will call `io.resume_write` that will wake only one single commit.

If the `action_id` were a Fiber, then it would wake up that fiber upon completion. It is more suitable for other operations like read/write/connect/send/etc. Currently only we only do poll with a callback.